clients/src/main/java/org/apache/kafka/common/requests/LeaveGroupResponse.java

```diff
@@ -17,47 +17,126 @@
+                                                    "response, while actually get " + memberResponses.size());
+            }
+
+            short errorCode = topLevelError != Errors.NONE ? topLevelError.code() :
+                                  memberResponses.get(0).errorCode();
+
+            this.data = new LeaveGroupResponseData()
+                            .setErrorCode(errorCode);
+        } else {
+            this.data = new LeaveGroupResponseData()
+                            .setErrorCode(topLevelError.code())
+                            .setMembers(memberResponses);
+        }
+
+        if (version >= 1) {
+            this.data.setThrottleTimeMs(throttleTimeMs);
+        }
+    }
+
     public LeaveGroupResponse(Struct struct) {
         short latestVersion = (short) (LeaveGroupResponseData.SCHEMAS.length - 1);
         this.data = new LeaveGroupResponseData(struct, latestVersion);
     }
+
     public LeaveGroupResponse(Struct struct, short version) {
         this.data = new LeaveGroupResponseData(struct, version);
     }
 
-    public LeaveGroupResponseData data() {
-        return data;
-    }
-
     @Override
     public int throttleTimeMs() {
         return data.throttleTimeMs();
     }
 
+    public List<MemberResponse> memberResponses() {
+        if (data.members().isEmpty()) {
+            return Collections.singletonList(
+                new MemberResponse().setErrorCode(data.errorCode())
+            );
+        } else {
+            return data.members();
+        }
+    }
+
     public Errors error() {
-        return Errors.forCode(data.errorCode());
+        Errors topLevelError = Errors.forCode(data.errorCode());
+        if (topLevelError != Errors.NONE) {
+            return topLevelError;
+        } else {
+            for (MemberResponse memberResponse : data.members()) {
+                Errors memberError = Errors.forCode(memberResponse.errorCode());
+                if (memberError != Errors.NONE) {
+                    return memberError;
+                }
+            }
+            return Errors.NONE;
+        }
     }
 
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return Collections.singletonMap(Errors.forCode(data.errorCode()), 1);
+        Map<Errors, Integer> combinedErrorCounts = new HashMap<>();
+        // Top level error.
+        Errors topLevelError = Errors.forCode(data.errorCode());
+        if (topLevelError != Errors.NONE) {
+            combinedErrorCounts.put(topLevelError, 1);
+        }
+
+        // Member level error.
+        Map<String, Errors> errorMap = new HashMap<>();
+
+        for (MemberResponse memberResponse : data.members()) {
+            Errors memberError = Errors.forCode(memberResponse.errorCode());
+            if (memberError != Errors.NONE) {
+                errorMap.put(memberResponse.memberId(), memberError);
+            }
+        }
+
+        combinedErrorCounts.putAll(errorCounts(errorMap));
```

Comment:
The recursive call seems like overkill. It seems like we could just create a single collection and use `AbstractResponse.updateErrorCounts`.